### PR TITLE
Enable local-up-cluster.sh to start with the secured port

### LINKED ISSUE #33375: Extend `local-up-cluster.sh` to be able to start with the secured port
Followup to https://github.com/kubernetes/kubernetes/pull/31491

Extend `local-up-cluster.sh` to be able to start with the secured port available using https://github.com/kubernetes/kubernetes/pull/31491 and an option to close the insecured port so that its easy to start determining which permissions the controllers need.

The controller manager should try to make use of the secured port hidden behind an option since that's broken at the moment.

### PR DESCRIPTION
**Pull request by dims**
related to #33375

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33966)

### CODE REVIEW

File: hack/local-up-cluster.sh [hunk 1]
```diff
@@ -139,15 +140,18 @@ CPU_CFS_QUOTA=${CPU_CFS_QUOTA:-true}
...
 ENABLE_CONTROLLER_ATTACH_DETACH=${ENABLE_CONTROLLER_ATTACH_DETACH:-"true"} # current default
+CERT_DIR=${CERT_DIR:-"/var/run/kubernetes"}
+ROOT_CA_FILE=$CERT_DIR/apiserver.crt
+
 
 function test_apiserver_off {
     # For the common local scenario, fail fast if server is already running.
     # this can happen if you run local-up-cluster.sh twice and kill etcd in between.
-    curl --silent -g $API_HOST:$API_PORT
+    curl --silent -k -g $API_HOST:$API_SECURE_PORT
```
Comment (eparis) [inline hack/local-up-cluster.sh]: Why not use `--cacert=$(ROOT_CA_FILE)` instead of `-k` ?

File: hack/local-up-cluster.sh [hunk 2]
```diff
@@ -545,8 +568,9 @@ To start using your cluster, open up another terminal/tab and run:
 
   export KUBERNETES_PROVIDER=local
 
-  cluster/kubectl.sh config set-cluster local --server=http://${API_HOST}:${API_PORT} --insecure-skip-tls-verify=true
-  cluster/kubectl.sh config set-context local --cluster=local
+  cluster/kubectl.sh config set-cluster local --server=https://${API_HOST}:${API_SECURE_PORT} --insecure-skip-tls-verify=true
```
Comment (eparis) [inline hack/local-up-cluster.sh]: do we still need the ``insecure-skip-tls-verify` ?

Comment (dims) [inline hack/local-up-cluster.sh]: yes, cleaning up now

**Comment by dims** [comment]
@k8s-bot kubemark e2e test this

**Comment by dims** [comment]
@k8s-bot gci gce e2e test this

**Comment by dims** [comment]
ok, here's what i got working so far:

1. Run hack/local-up-cluster.sh, then stop it with CTRL-C
2. disable the insecure port using `export API_PORT=0`
3. Run hack/local-up-cluster.sh

Essentially apiserver.key/apiserver.crt are generated in step #1 and the local client can use the apiserver.crt. 

Note that we create the local client first and then generate the certs when we start the secure listener. Hence the two step tango. Any ideas to make this better? Please let me know!

**Comment by dims** [comment]
@deads2k : your thoughts please? (See my last comment)

**Comment by deads2k** [comment]
I think that this will order the parameters in a way that allows it to work on the first try: https://github.com/kubernetes/kubernetes/pull/34556 .

**Comment by dims** [comment]
We need #34556 to merge first

**Comment by deads2k** [comment]
@dims I definitely want this, but I'm not sure it completely fixes #33375 since the `kube-controller-manager` won't have an identity, right?

This pull should still merge before that (it's a big step forward), but I don't think we're all the way there yet.

**Comment by dims** [comment]
@deads2k yep, indeed. we can leave the #33375 open and leave notes as to what more steps are needed.

File: hack/local-up-cluster.sh [hunk 3]
```diff
@@ -139,15 +140,18 @@ CPU_CFS_QUOTA=${CPU_CFS_QUOTA:-true}
...
+
 
 function test_apiserver_off {
     # For the common local scenario, fail fast if server is already running.
     # this can happen if you run local-up-cluster.sh twice and kill etcd in between.
-    curl --silent -g $API_HOST:$API_PORT
+    curl --silent -k -g $API_HOST:$API_SECURE_PORT
     if [ ! $? -eq 0 ]; then
-        echo "API SERVER port is free, proceeding..."
+        echo "API SERVER secure port is free, proceeding..."
```
Comment (deads2k) [inline hack/local-up-cluster.sh]: probably ought to check both.

Comment (dims) [inline hack/local-up-cluster.sh]: Ack. Fixing

File: pkg/genericapiserver/options/server_run_options.go [hunk 4]
```diff
@@ -227,8 +228,21 @@ func (s *ServerRunOptions) NewSelfClientConfig(token string) (*restclient.Config
 		QPS:   50,
 		Burst: 100,
 	}
-	if s.SecurePort > 0 && len(s.TLSCAFile) > 0 {
-		clientConfig.Host = "https://" + net.JoinHostPort(s.BindAddress.String(), strconv.Itoa(s.SecurePort))
+
+	// Use secure port if the TLSCAFile is specified and the file exists
```
Comment (dims) [inline pkg/genericapiserver/options/server_run_options.go]: Right, removing the fallback to insecure

**Comment by dims** [comment]
@eparis @deads2k Addressed all the comments and tested `API_PORT=0` locally after cleaning up all artifacts (`sudo rm -rf /var/run/kubernetes/ /tmp/kube* ~/.kube/`)

**Comment by deads2k** [comment]
`API_PORT=0 hack/local-up-cluster.sh` is ending in `!!! [1014 10:36:12] Timed out waiting for apiserver:  to answer at https://localhost:6443/api/v1/pods; tried 10 waiting 1 between each
Cleaning up...
` for me

**Comment by dims** [comment]
@deads2k can you paste me your /tmp/kube-apiserver.log please?

**Comment by deads2k** [comment]
> @deads2k can you paste me your /tmp/kube-apiserver.log please?

https://gist.github.com/deads2k/6181c16615618daeb7b3d6fdb8e16f72

**Comment by dims** [comment]
@deads2k @sttts - found a crumb to solve the mystery. seems curl behaves differently with the certs we generate when using `--cert-dir`

```
[fedora@my-fedora kubernetes]$ curl -v --cacert /var/run/kubernetes/apiserver.crt -gfs https://localhost:6443/api/v1/pods
*   Trying ::1...
* Connected to localhost (::1) port 6443 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
*   CAfile: /var/run/kubernetes/apiserver.crt
  CApath: none
* Server certificate:
* 	subject: CN=149.202.163.230@1476827403
* 	start date: Oct 18 21:50:03 2016 GMT
* 	expire date: Oct 18 21:50:03 2017 GMT
* 	common name: 149.202.163.230@1476827403
* 	issuer: CN=149.202.163.230@1476827403
* NSS error -8101 (SEC_ERROR_INADEQUATE_CERT_TYPE)
* Certificate type not approved for application.
* Closing connection 0
```

**Comment by dims** [comment]
So for now i'll have to switch `hack/lib/util.sh` to just use `-k` instead of `--ca-cert`

**Comment by sttts** [comment]
@dims the cert error with curl comes from the "X509v3 Extended Key Usage" field in the cert. Certs with this attribute (which includes the usage as server cert in our case) cannot be used as CA with curl (and Firefox afaik). The solution is to create a root CA with one key and another key for the actual apiserver cert. This would be much cleaner IMO.

**Comment by dims** [comment]
@sttts are we talking about fixing --cert-dir automatic generation of certs (apiserver.key/apiserver.crt). Adding an additional root CA file there?

Or are we talking about using say cfssl to generate stuff so we can pass in --tls-* flags (instead of using --cert-dir)?

Either way, can we do that in next/another PR?

Thanks,
Dims

**Comment by sttts** [comment]
@dims my guess is that the auto generated cert is broken for curl under certain circumstances. Of course, using cfssl would be the way to go for a rootca+apiserver.crt implementation, but this can wait for another PR.

I tested your changes on my machine with success. Still wondering why it doesn't work for @deads2k. Am on Fedora 24 here.

I saw that curl error before when I had a broken API_HOST=0.0.0.0 in my environment. While troubleshooting I was able to make curl happy with https://gist.github.com/sttts/b45105abe7779f89aa75ef90638a48c7.

**Comment by deads2k** [comment]
If it works for @sttts I'll merge as is and hack on it later.

**Comment by liggitt** [comment]
the auto-generated cert is a CA *and* a server cert

**Comment by dims** [comment]
@liggitt agree. it's just that it does not work well on fedora. One option seems to be that we should do as @sttts suggests (https://gist.github.com/sttts/b45105abe7779f89aa75ef90638a48c7)

**Comment by sttts** [comment]
@liggitt @dims did a rough cfssl implementation with root-ca today: https://github.com/kubernetes/kubernetes/pull/35143. Curl and go's tls stack are happy.

**Comment by dims** [comment]
@sttts very cool. that was fast 👍

**Pull request merged**